project/src/main/ui/chat/chatscript-parser.gd

```diff
@@ -0,0 +1,378 @@
+class_name ChatscriptParser
+"""
+Parsers a chatscript (.chat) file containing cutscene or dialog data.
+
+This class is stateful and intended to be used once and thrown away. If it is reused, the previously parsed chat_tree
+will be erased.
+"""
+
+"""
+Tracks the state for the parser.
+
+The ChatscriptParser has different states based on whether it's parsing dialog, characters, locations, or something
+else.
+"""
+class AbstractState:
+	var chat_tree: ChatTree
+	
+	func _init(init_chat_tree: ChatTree) -> void:
+		chat_tree = init_chat_tree
+	
+	
+	func line(_line: String) -> String:
+		return ""
+
+
+"""
+Default parser state for parsing headers and metadata.
+
+Mostly contains logic for transitioning into other states.
+"""
+class DefaultState extends AbstractState:
+	var dialog_state: AbstractState
+	
+	func _init(init_chat_tree: ChatTree).(init_chat_tree) -> void:
+		pass
+	
+	
+	func line(line: String) -> String:
+		var result := ""
+		if not line:
+			pass
+		elif line.begins_with("{") and line.ends_with("}"):
+			# json dictionary; set metadata
+			var parsed = parse_json(line)
+			if typeof(parsed) == TYPE_DICTIONARY:
+				chat_tree.meta = parse_json(line)
+				if "version" in chat_tree.meta:
+					chat_tree.meta.erase("version")
+			else:
+				push_warning("Malformed chatscript JSON: %s" % [line])
+		elif line.begins_with("[") and line.ends_with("]"):
+			# state name; update state
+			result = line.trim_prefix("[").trim_suffix("]")
+		else:
+			# unrecognized text; parse as dialog and change state
+			result = dialog_state.line(line)
+			result = StringUtils.default_if_empty(result, DIALOG)
+		return result
+
+
+"""
+Parser state for parsing location data (where a conversation takes place).
+"""
+class LocationState extends AbstractState:
+	func _init(init_chat_tree: ChatTree).(init_chat_tree) -> void:
+			pass
+	
+	
+	"""
+	Syntax:
+		[location]
+		indoors
+	"""
+	func line(line: String) -> String:
+		var result := ""
+		if line:
+			chat_tree.location_id = line
+		else:
+			result = DEFAULT
+		return result
+
+
+"""
+Parser state for parsing character data (participants in a conversation).
+"""
+class CharactersState extends AbstractState:
+	var _character_aliases: Dictionary
```

Comment:
Document this field

Reply:
Added comment, `Creature aliases used in chatscript dialog. DialogState references this dictionary when parsing dialog`